server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveClusterAction.java

```diff
@@ -0,0 +1,455 @@
+ *
+ * To determine whether any indices, aliases or datastreams match the index expression passed in
+ * on each cluster, the ResolveIndexAction functionality is used. ResolveIndexAction provides a
+ * complete listing of all matching indices, aliases or datastreams, whereas this endpoint only
+ * provides a boolean value about whether anything matches at all. For detailed information
+ * about which indices, aliases and/or datastreams match the _resolve/index endpoint should be used.
+ */
+public class ResolveClusterAction extends ActionType<ResolveClusterAction.Response> {
+
+    public static final ResolveClusterAction INSTANCE = new ResolveClusterAction();
+    public static final String NAME = "indices:admin/resolve/cluster";
+
+    public static class Request extends ActionRequest implements IndicesRequest.Replaceable {
+
+        // we only allow querying against open, non-hidden indices
+        public static final IndicesOptions DEFAULT_INDICES_OPTIONS = IndicesOptions.strictExpandOpen();
+
+        private String[] names;
+        private IndicesOptions indicesOptions = DEFAULT_INDICES_OPTIONS;
+
+        public Request(String[] names) {
+            this.names = names;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.names = in.readStringArray();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeStringArray(names);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ResolveClusterAction.Request request = (ResolveClusterAction.Request) o;
+            return Arrays.equals(names, request.names);
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(names);
+        }
+
+        @Override
+        public String[] indices() {
+            return names;
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return indicesOptions;
+        }
+
+        @Override
+        public IndicesRequest indices(String... indices) {
+            this.names = indices;
+            return this;
+        }
+
+        @Override
+        public boolean allowsRemoteIndices() {
+            return true;
+        }
+
+        @Override
+        public boolean includeDataStreams() {
+            // request must allow data streams because the index name expression resolver for the action handler assumes it
+            return true;
+        }
+    }
+
+    private ResolveClusterAction() {
+        super(NAME, ResolveClusterAction.Response::new);
+    }
+
+    public static class ResolveClusterInfo implements Writeable {
```

Comment:
Thanks for raising. I have a question out to core-infra.

Reply:
I don't think we've done new TransportVersions for _new_ writeables, only for changing to existing ones. If the new action is sent to an old node, the old node would just reject the request as unknown, right? I see 3 cases:

* New node sends action to old node - error, regardless of new transport version
* New node sends action to current node - ok
* Old node gets request - error, it doesn't know about the request

I think in these cases we just have to accept that you can't reasonably send a new request type/action while the cluster is being upgraded.

Reply:
++ that makes sense to me

Reply:
Addendum: at the very least I think we should throw an exception while serializing the request to a transport version which doesn't understand it. No need to round-trip nonsense over the wire.

Reply:
> we should be rejecting actions that the cluster doesn't fully support by checking the min cluster transport version or the cluster features

I agree with this. I guess it's a question of *how* we reject the requests. Checking min transport version or feature is a way to do it before hitting the node the request will go to. Ultimately, though, the request is still going to fail in either case.

Reply:
> Ultimately, though, the request is still going to fail in either case.

Checking the min transport version is not quite the same as sending out the requests and letting the remote nodes fail if they don't understand them. Firstly we can give clearer feedback up front if the min transport version is insufficient. But also if we are two-thirds of the way through upgrading a three-node cluster then it is IMO preferable if all new-version-only requests fail, even ones between the two upgraded nodes. It's just that much simpler for users to have features that don't work at all until the upgrade is complete, rather than a feature that sort of limps into existence as the upgrade progresses.

Reply:
That's a good point. I agree we should fail early, and consistently.

The remaining question for me is whether every new action should request a new transport version/feature, or whether it should be automatic, based on knowing what actions the node/cluster is capable of. That's a future idea though, so in this case, we should move forward with a new transport version or feature for now.

Reply:
I finished responding to most of the other comments and now am ready to tackle this one. I believe we are saying that I should add a new TransportVersion and wrap all the new Writeables with an if guard.

What do I do if the version is below the allowed threshold?
Is there an example of doing this that I can follow, or is this a new policy?

I tried an [initial commit here](https://github.com/elastic/elasticsearch/pull/102726/commits/31a741dd702fdd8eccd66dc54f36ced16d2f9099) with an if-guard where I throw `UnsupportedOperationException`. Let me know if that's what you are thinking we should do. (I probably also need to add this to the `ResolveClusterInfo` class.)

Reply:
An additional consideration - because cross-cluster version support currently goes back only one minor (e.g., 8.13 can back to to 8.12), Product asked me whether I can backport this endpoint to the previous minor branch. Adding a new TransportVersion here and if-guard around the Writeables would make that impossible, right? 

A key reason for this new endpoint is for Kibana to determine whether remote clusters are connected to the querying cluster and provide other metadata to determine if the user wants to send the CCS request there.

Before adding the above if-guard, I already have code to handle the case where the remote cluster does not have the endpoint: https://github.com/elastic/elasticsearch/pull/102726/files#diff-7f31fad49d2c76462e47c62de33d00c20676978c866d78fe67cbeacde2410954R212

I like that model because it still allowed me to tell the user that the cluster is in fact connected to the cluster even though it is an older version. In that case, you would get back in the response:

```
  "remote1": {
    "connected": true,
    "skip_unavailable": false
  }
```

(The `matching_indices` and `version` entries are not there because the cluster doesn't have the code to return the response, but I can tell it's connected).

With the new model using if/guards around the Writeables, I get:

```
  "remote1": {
    "connected": false,
    "skip_unavailable": false,
    "error": "java.lang.UnsupportedOperationException: ResolveClusterAction requires at least Transport Version 8563000 but was 8500061"
  }
```

The logs show:
```
Caused by: java.lang.UnsupportedOperationException: ResolveClusterAction requires at least Transport Version 8563000 but was 8500061
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.action.admin.indices.resolve.ResolveClusterActionRequest.writeTo(ResolveClusterActionRequest.java:52)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.OutboundMessage.serialize(OutboundMessage.java:71)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.OutboundHandler.sendMessage(OutboundHandler.java:185)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.OutboundHandler.sendRequest(OutboundHandler.java:105)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.TcpTransport$NodeChannels.sendRequest(TcpTransport.java:321)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.RemoteConnectionManager$InternalRemoteConnection.sendRequest(RemoteConnectionManager.java:355)
        at org.elasticsearch.server@8.13.0-SNAPSHOT/org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:952)
```

So now we incorrectly state that `connected=false` for this cluster.

Which leads back to the question of how do we do a backport to the previous minor to avoid this issue?

Or other ideas to solve this dilemma?

Reply:
Throwing some kind of exception is about the best we can do today. I'm unsure of the process for backporting this sort of thing - if it's even possible it certainly won't be easy.

Reply:
> Before adding the above if-guard, I already have code to handle the case where the remote cluster does not have the endpoint

Yes that makes sense, but you can do the same thing with a UOE thrown locally too. And then fall back to some other (valid) cross-cluster request to ensure we're connected and to get whatever info we can without the new action.

Reply:
OK, thanks. I'll work on that next.

Reply:
@dct - how would you recommend that I call the ResolveIndexAction endpoint from the ResolveClusterAction callback. I have an (incorrect) example in the latest push: https://github.com/elastic/elasticsearch/pull/102726/commits/e7aee512da50fc78005a1f695f6d66d13bac5798#diff-7f31fad49d2c76462e47c62de33d00c20676978c866d78fe67cbeacde2410954R190

That call "works" in the sense that the call to the remote cluster works (I see it in the logs), but of course, that secondary ActionListener callback is not chained to the original TransportResolveClusterAction callback. Is there an example of this elsewhere in the codebase I can copy? I'm looking at helpers in ActionListener presently and maybe I should use `runAfter` or `completeWith`?

Thanks for your help on how to do this.

Reply:
I think you just want another `ActionListener.releaseAfter(..., refs.acquire())`

Reply:
Yep, that works, thanks. Will be in next push.